File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = 0;
        
        /**
        *   플로이드 워셜을 활용해 각 지점 간의 최단거리를 계산한다.
        *       - n의 최대값은 200이므로, O(n^3)의 알고리즘도 충분히 활용할 만 하다.
        *   S에서 특정 지점까지의 거리 + 해당 지점에서 각각 A, B까지의 거리를 합한 값의 최소값을 반환한다.
        */
        int[][] cost = new int[n + 1][n + 1];
        
        for (int i = 1; i <= n; i++) {
            Arrays.fill(cost[i], 20000002);
            cost[i][i] = 0;
        }
        
        for (int[] fare: fares) {
            cost[fare[0]][fare[1]] = fare[2];
            cost[fare[1]][fare[0]] = fare[2];
        }

        floydWarshall(cost, n);
        
        answer = Integer.MAX_VALUE;
        for (int i = 1; i <= n; i++) {
            answer = Math.min(cost[s][i] + cost[i][a] + cost[i][b], answer);
        }
        
        return answer;
    }
    
    void floydWarshall(int[][] cost, int n) {
        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    if (cost[i][j] > cost[i][k] + cost[k][j]) {
                        cost[i][j] = cost[i][k] + cost[k][j];
                    }
                }
            }
        }
    }
}

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = INT_MAX;
    /**
    *   S에서 각각 A, B로 향할 때, 최소 비용으로 연결하는 것은
    *   A(혹은 B)에서 S를 지나 B(혹은 A)에 도착하는 최단경로를 계산하는 것과 동일하다.
    *   따라서 플로이드 워셜을 통해, 각 지점에서 다른 모든 지점까지의 최단거리를 계산한다.
    *   이후 모든 지점 X에 대해 A-X, S-X, B-X 를 더한 값의 최솟값을 반환한다.
    */
    
    // i에서 j로 c의 비용의 간선을 가지는 인접배열
    vector<vector<int>> map (n + 1, vector<int>(n + 1, INT_MAX));
    
    for (int i = 1; i <= n; i++) {
        map[i][i] = 0;
    }
    
    for (auto& fare: fares) {
        int x = fare[0], y = fare[1], cost = fare[2];
        
        map[x][y] = cost;
        map[y][x] = cost;
    }
    
    for (int k = 1; k <= n; k++) {
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (map[i][k] != INT_MAX && map[k][j] != INT_MAX && map[i][k] + map[k][j] < map[i][j]) {
                    map[i][j] = map[i][k] + map[k][j];
                }
            }
        }
    }
    
    for (int i = 1; i <= n; i++) {
        answer = min(map[a][i] + map[s][i] + map[b][i], answer);
    }
    return answer;
}